src/Parser/Core/Modules/CastEfficiency.js

```diff
@@ -4,182 +4,210 @@ import SpellLink from 'common/SpellLink';
+        } else if (averageCooldown) { // no average CD if spell hasn't been cast
+          rawMaxCasts = (this.owner.fightDuration / averageCooldown) + (ability.charges || 1) - 1;
+        } else {
+          rawMaxCasts = (this.owner.fightDuration / cooldownMs) + (ability.charges || 1) - 1;
+        }
+        const maxCasts = Math.ceil(rawMaxCasts) || 0;
+        const maxCpm = (cooldown === null) ? null : maxCasts / fightDurationMinutes;
+
+        let castEfficiency;
+        if(ability.getMaxCasts) { // legacy support for custom getMaxCasts
+          castEfficiency = Math.min(1, casts / rawMaxCasts);
+        } else {
+          // Cast efficiency calculated as the percent of fight time spell was on cooldown
+          const timeOnCd = this._getTimeOnCooldown(spellId);
+          castEfficiency = (timeOnCd / this.owner.fightDuration) || null;
+        }
+
+        const recommendedCastEfficiency = ability.recommendedCastEfficiency || DEFAULT_MINOR_SUGGEST;
+        const averageIssueCastEfficiency = ability.averageIssueCastEfficiency || DEFAULT_AVERAGE_SUGGEST;
+        const majorIssueCastEfficiency = ability.majorIssueCastEfficiency || DEFAULT_MAJOR_SUGGEST;
+
+        const canBeImproved = castEfficiency !== null && castEfficiency < recommendedCastEfficiency && casts < maxCasts;
+
+        return {
+          ability,
+          cpm,
+          maxCpm,
+          casts,
+          maxCasts,
+          castEfficiency,
+          recommendedCastEfficiency,
+          averageIssueCastEfficiency,
+          majorIssueCastEfficiency,
+          canBeImproved,
+        };
+      })
+      .filter(item => item !== null);
   }
 
   suggestions(when) {
-    const castEfficiency = getCastEfficiency(this.constructor.CPM_ABILITIES, this.abilityTracker, this.combatants, this.owner);
-    castEfficiency.forEach(cpm => {
-      if (cpm.ability.noSuggestion || cpm.castEfficiency === null) {
+    const castEfficiencyInfo = this._generateCastEfficiencyInfo();
+    castEfficiencyInfo.forEach(abilityInfo => {
+      if(abilityInfo.ability.noSuggestion || abilityInfo.castEfficiency === null || abilityInfo.casts === abilityInfo.maxCasts) {
         return;
       }
-      when(cpm.castEfficiency).isLessThan(cpm.recommendedCastEfficiency)
+      const ability = abilityInfo.ability;
+      when(abilityInfo.castEfficiency).isLessThan(abilityInfo.recommendedCastEfficiency)
         .addSuggestion((suggest, actual, recommended) => {
-          return suggest(<span>Try to cast <SpellLink id={cpm.ability.spell.id} /> more often. {cpm.ability.extraSuggestion || ''} <a href="#spell-timeline">View timeline</a>.</span>)
-            .icon(cpm.ability.spell.icon)
-            .actual(`${cpm.casts} out of ${cpm.maxCasts} possible casts; ${formatPercentage(actual)}% cast efficiency`)
+          return suggest(<span>Try to cast <SpellLink id={ability.spell.id} /> more often. {ability.extraSuggestion || ''} <a href="#spell-timeline">View timeline</a>.</span>)
```

Comment:
Using the Wrapper component in `common/Wrapper` is recommended instead of spans to reduce the amount of unnecessary elements being rendered.

Reply:
I assume that means <Wrapper> instead of <span>. I don't know enough about html to understand why it makes a difference, but I'll do it.

Reply:
Wrapper doesn't render anything, span renders span

Reply:
I don't know what that means 😆

Reply:
it's fake https://github.com/WoWAnalyzer/WoWAnalyzer/blob/master/src/common/Wrapper.js

Reply:
I understand the Wrapper is like a "fake object", I just don't understand what it means to 'render span', and why it's bad that we're rendering span.

Reply:
The more elements we render the slower things get, see Events tab. A bunch of spans won't matter but doing this properly everywhere will improve render performance.

Reply:
Rendering is hundreds of times slower than just running JS.